File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        d = ListNode(0 , head)
        prev, curr = d, head
        while curr and curr.next:
            nxtPair = curr.next.next
            second = curr.next

            second.next = curr
            curr.next = nxtPair
            prev.next = second

            prev = curr 
            curr = nxtPair
        return d.next

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode *dummy = new ListNode(0, head);
        ListNode *prev = dummy;
        // ListNode **curr = &head;
        while (head && head->next)
        {
            ListNode *nxtPair = head->next->next;
            ListNode *second = head->next;

            second->next = head;
            head->next = nxtPair;
            prev->next = second;

            prev = head;
            head = nxtPair;

        }
        return dummy->next;
    }
};